src/vm/circuit.rs

```diff
@@ -0,0 +1,333 @@
+    let [add_mod @ Value::Unit, mul_mod @ Value::Unit, _descripctor @ Value::Unit, Value::Circuit(inputs), Value::CircuitModulus(modulus), _, _]: [Value; 7] = _args.try_into().unwrap()
+    else {
+        panic!()
+    };
+    let circ_info = match _registry.get_type(&_info.ty).unwrap() {
+        CoreTypeConcrete::Circuit(CircuitTypeConcrete::Circuit(info)) => &info.circuit_info,
+        _ => todo!(),
+    };
+    let mut outputs = inputs
+        .into_iter()
+        .enumerate()
+        .map(|(i, input)| {
+            let gate_num = &circ_info.mul_offsets[i];
+            (gate_num.output as u64, input)
+        })
+        .collect::<HashMap<u64, BigUint>>();
+
+    let mut add_gates = circ_info.add_offsets.iter().peekable();
+    let mut mul_gates = circ_info
+        .mul_offsets
+        .iter()
+        .skip(circ_info.n_inputs)
+        .peekable();
+
+    let success = loop {
+        while let Some(add_gate) = add_gates.peek() {
+            let lhs = outputs.get(&(add_gate.lhs as u64));
+            let rhs = outputs.get(&(add_gate.rhs as u64));
+
+            match (lhs, rhs) {
+                (Some(l), Some(r)) => {
+                    outputs.insert(add_gate.output as u64, (l + r) % &modulus);
+                }
+                (None, Some(r)) => {
+                    let res = match outputs.get(&(add_gate.output as u64)) {
+                        Some(res) => res,
+                        None => break,
+                    };
+                    // if it is a sub_gate the output index is store in lhs
+                    outputs.insert(add_gate.lhs as u64, (res + &modulus - r) % &modulus);
+                }
+                // there aren't enough gates computed for add_gate to compute
+                // the next gate so we need to compute a mul_gate
+                _ => break,
+            };
+
+            add_gates.next();
+        }
+
+        match mul_gates.next() {
+            Some(mul_gate) => {
+                let lhs = outputs.get(&(mul_gate.lhs as u64));
+                let rhs = outputs.get(&(mul_gate.rhs as u64));
+
+                match (lhs, rhs) {
+                    (Some(l), Some(r)) => {
+                        outputs.insert(mul_gate.output as u64, (l * r) % &modulus);
+                    }
+                    (None, Some(r)) => {
+                        let res = match r.modinv(&modulus) {
+                            Some(inv) => inv,
+                            None => {
+                                // attempted to get the inverse of 0,
+                                break false;
+                            }
+                        };
+                        // if it is a inv_gate the output index is store in lhs
+                        outputs.insert(mul_gate.lhs as u64, res);
+                    }
+                    // a mul_gate can always be computed because it is only computed
+                    // if an add_gate can't
+                    _ => continue,
```

Comment:
we should panic or return here as an invalid circuit would end up in an infinite loop